test/util/listener/listener.go

```diff
@@ -19,6 +20,7 @@ func (addr) String() string  { return "testlistener" }
 type Listener struct {
 	c      chan net.Conn
 	closed bool
+	sync.Mutex
```

Comment:
I would introduce sync.Once here, e.g.
```suggestion
	closeOnce sync.Once
```

Reply:
Thanks for the review @yehornaumenko. 

I am not sure if that would work. Looking at the sync.Once [documentation](https://pkg.go.dev/sync#Once), it says that the action will be performed just once (in total, not "concurrently"), so the behavior of the code will be different as we want it to be executed by a maximum of one goroutine at an exact given time, not in total. 

If we also take a look at the .Do [documentation](https://pkg.go.dev/sync#Once.Do), we see that it says 
_Do calls the function f if and only if Do is being called for the first time for this instance of Once_, 
so we can confirm this is different than what we want.

It is also interesting to look at the Go [source](https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/sync/once.go;l=69;drc=03f6d81fc7d52ec53deb94cff69b63d04e689c24;bpv=0;bpt=1) code to understand how the behavior is different. 

Rejecting the suggestion but an interesting comparison.
😃